### ISSUE TITLE
zpp fail on overflow

### ISSUE THREAD

**Issue by jwatzman** [facebook/hhvm#6048]
https://wiki.php.net/rfc/zpp_fail_on_overflow

We can probably just do this; it's a BC break, but the old behavior was pretty dangerous.

**Comment by lexidor** [comment]
@jwatzman There is no repro in the php rfc, and they only mention 32bit. HHVM only runs on 64bit platforms. Is hhvm (still) affected by this?

**Comment by jwatzman** [comment]
- I am pretty sure that `double` (i.e., 64-bit floating point numbers) has a maximum value well above that of 64-bit integers? And I assume that's the representation used on 64-bit platforms? (Not 100% sure of either of those!) So this may still be relevant on 64-bit.
- That said, this issue was filed when I/we were working on PHP7 support in HHVM. I think I just went through all of the PHP7 RFCs and filed issues on ones that looked relevant. With HHVM dropping PHP support altogether, I'm not sure what the current thinking is on stuff like this?
- On the other other hand, this is still probably a really good idea to do anyway from a safety perspective -- if HHVM's argument parsing hasn't changed enough that it's already doing this! (It might have even been back in 2015, I don't think I ever actually checked?)

Basically... I think this can happen even on 64-bit, and although it was obstensibly for PHP7 support, it's probably a good idea for safety for HHVM to implement this anyway, if it doesn't already? Just my opinion though -- I haven't been on the HHVM team in 4.5 years :)

**Comment by lexidor** [comment]
> I haven't been on the HHVM team in 4.5 years :)
Sorry, I was not aware of this. I will not `@` you on other issues.

Yes this biggest float is indeed greater than 2**63. Doubles reach into the 2**300s iirc.

This rfc talks about passing a big float to a function expecting an int. I can't find a good way to repro this, since I'd need to find a stdlib function that takes a `mixed`, but expects an int. If it takes an `int`, this will throw a `\TypeError` since hhvm 4.3.0.

**Comment by jwatzman** [comment]
> Sorry, I was not aware of this. I will not @ you on other issues.

No worries! Happy to help out where I can offer historical context. Just pointing out that you shouldn't take anything I say as normative or official :-P

(As an aside, thanks for going through all these old issues!)

> If it takes an int, this will throw a \TypeError since hhvm 4.3.0.

Then if I understand, this may no longer be relevant to HHVM?

**Comment by lexidor** [comment]
> Then if I understand, this may no longer be relevant to HHVM?

Well that is the hard part. ZendParamModeCoerce is dead, since all built-ins now throw on incorrect types. So, if this was indeed something that was done wrong systematically by the translation from VM land to cpp land, we're fine.

However, there could be functions that take a cpp type of `&Variant` (`mixed` in the ext_herpderp.php files) and then do this incorrect conversion in the body of the cpp function. I don't know if this is the case and I can't systematically go over every built-in function to check. 

However, I don't think this will be that much of an issue. `Variant` has many methods on it like `toCppString()`, `toInt64()`, and friends. As long as these methods are well behaved, I think I can close this issue. Any method that does this conversion wrong on its own, is a bug with that method, not with hhvm as a whole.

Hhvm gives me direct access to `Variant->toInt64()` via the `__PHPStdLib` function [`\intval()`](https://github.com/facebook/hhvm/blob/3b45b370cd2b821c6ebdd36d1cf3f70e44065e89/hphp/runtime/ext/std/ext_std_variable.cpp#L85). I'll throw some floats at that to try and break it. PHP specifies that `intval(float_greater_than_php_int_max)` should return `\PHP_INT_MAX`.

> (As an aside, thanks for going through all these old issues!)

Thank you, I asked Fred if I could be helpful this way, and he made this thing possible. Fred is amazing.

**Comment by lexidor** [comment]
Wait what? `Variant->toInt64()` actually flipped the sign!

What is going on here?

```HACK
<<__EntryPoint>>
function main(): void {
    $signed_int64_max_as_float = (float)\PHP_INT_MAX;
    if (
        $signed_int64_max_as_float == \PHP_INT_MAX &&
        $signed_int64_max_as_float > 0
    ) {
        echo "Float cast successful, it remained signed int64 max\n";
    }

    \printf("%f\n", $signed_int64_max_as_float);

    $as_int = \intval($signed_int64_max_as_float);
    \printf("%d\n", $as_int);

    if ($as_int < 0) {
        echo "ERROR: big float became a negative int.\n";
    }
}
```

```
Float cast successful, it remained signed int64 max
9223372036854775808.000000
-9223372036854775808
ERROR: big float became a negative int.
```

**Comment by lexidor** [comment]
Oh, I see what happens here. `\PHP_INT_MAX` is `...807`. The float is `...808`. `\PHP_MAX_INT` + 1 rolls over to `\PHP_MIN_INT`.
